### ISSUE TITLE
ft_prepare_sourcemodel: improve documentation/functionality of inwardshift/moveinward

### ISSUE THREAD

**Issue by schoffelen** [fieldtrip/fieldtrip#1333]
Reported by @britta-wstnr by e-mail:

Documenting and taking it from here

1) we just came across some unexpected behavior (helping a student today) with ft_prepare_sourcemodel(). If we warp an MNI template, inwardshift has no effect - is that because that would shift grid points to the wrong positions after warping?

2) We often see problems with not inwardshifted sources when using the openmeeg pipeline for leadfield computation (either things breaking during leadfield computation or voxels that draw a lot of activity (several orders of magnitude higher than the rest) after source reconstruction). Fixing the template (i.e. changing what is inside/outside there) only helps in some cases. Actually, we saw yesterday that also setting cfg.moveinward but having that value being smaller than cfg.inwardshift introduces an inwardshift in the warping cases, however,  if cfg.moveinward > cfg.inwardshift it errors out.
That behavior seems a bit erratic to me, do you agree? I can check that this is true on another data set as well (so far it's only the data of the phd student, so maybe confirming with e.g. the FieldTrip example data set would be a good idea) unless you tell me this is expected behavior :)

**Comment by britta-wstnr** [comment]
I can have a look with the FieldTrip example data and the FieldTrip template later and report back.

**Comment by britta-wstnr** [comment]
I had a look with the FieldTrip `Subject 01` data and I see a similar behaviour pattern to what we saw with @jordannicolasalves:

Having specified `cfg.warpmni = 'yes'` (and supplying the template etc.) makes `cfg.inwardshift` have no effect (which might be wanted as it would destroy the warped MNI structure on an individual basis?).
If also `cfg.moveinward` is specified the grid points get definitely moved - unless `cfg.headmodel` is _not_ supplied, then the computation throws an error:
```
Undefined function or variable 'pos'.

Error in headsurface (line 208)
  ori = normals(pos, tri, 'vertex');

Error in ft_prepare_sourcemodel (line 683)
  [pos1, tri1] = headsurface(headmodel, [], 'inwardshift', cfg.moveinward, 'surface', 'brain');

Error in check_sourcem_inwardshift (line 77)
grid    = ft_prepare_sourcemodel(cfg);
```

If we agree that neither of the two options should be allowed with `cfg.warpmni`, I feel like this should be mentioned in the documentation and maybe a `warning` would be a good idea to give the user a heads-up.

I don't know whether the error is expected if the headmodel is not supplied.

**Comment by schoffelen** [comment]
note for reference: ft_prepare_sourcemodel has been revamped #1362, I would assume that the functionality hasn't changed, but the code logic should be more clear now. Just mention it here in case you are going to work on this function (looking forward to a PR!). To avoid conflicts you'd need to check out a recent version from the fieldtrip master branch.

**Comment by britta-wstnr** [comment]
@schoffelen I am happy to resolve some of the (my?) confusion around the `cfg.inwardshift`, e.g. by adding documentation or warnings/errors for undesirable combinations. I am, however, not really clear myself which combinations of `cfg.inwardshift`, `cfg.moveinward` and specific grids (warped or not) should be supported.

**Comment by schoffelen** [comment]
OK, I don't either. I think the first thing to do would be to get an idea about what you think should be supported functionality (i.e. ideally allowed combinations of cfg-arguments), and then evaluate how this compares to the current implementation. In case of discrepancies we then need to evaluate whether any missing functionality is missing 'by accident', or whether there's a deeper motivation to it. In the former situation, code might be adjusted to build in support for the missing functionality, in the second situation, appropriate user documentation should be written (in the help of the function) for proper user expectation management.

**Comment by robertoostenveld** [comment]
sorry for being slow to the party. These options are mainly the work of @vlitvak and me, and date quite some time back. Their functionality and interaction have changed a bit over time. 

Let me explain here, and then I hope you can make some sense of this, e.g. by extending the help documentation and/or a faq on the website. Note that I am doing it quickly from memory here, not by looking at the code.

A better name for `inwardshift` would be `insideshift`. It relates to shifting the volume conductor boundary that is used to detect whether dipoles are inside or outside. In the old days of a CTF localspheres  headmodel, the surface was the scalp and   it had to be shifted inward by 1,5 cm to mimic the brain surface.  But then we realized that having dipoles not too strictly in the brain but slightly outside helped when spatially normalizing, interpolating and averaging over subjects. And then we  switched to brain surface based head models, and then we ran into http://www.fieldtriptoolbox.org/faq/why_is_there_a_rim_around_the_brain_for_which_the_source_reconstruction_is_not_computed/ and then we figured that we could use a negative `inwardshift` to get better coverage of our interpolated group results. 

Another old-school use case for `inwardshift` is if you take a single sphere volume conductor, but a realistic head shape (e.g. measured with a polhemus) to constrain dipole positions. E.g. consider the initial grid search for dipole fitting a SEF, in case you only have the polhemus headshape but no individual MRI. The radius of the sphere is actually irrelevant, and the origin can be found by fitting it to the head surface. You still only want to consider dipole positions that are more or less inside the brain (so 1,5 cm inward from the scalp surface).  

The `moveinward` actually moves dipole positions. The application is when you take a template sourcemodel (e.g. the cortical sheet of colin27) and want to squeeze that into a non-matching headmodel. E.g. an individual head model, or even a sphere. I don't think it makes sense on a 3D grid, it would probably squeeze the complete cubic volume. 

For a source model based on a cortical sheet you don't want any vertex to be marked as outside, as that would give a "hole" in your estimation. So `inwardshift` should not apply (but probably the code is not aware of this situation, and happily will do its work).   

For a 3D grid source model you don't want `moveinward` to apply, since you dont want to squeeze the box into the brain (or into the sphere).

**Comment by britta-wstnr** [comment]
@robertoostenveld thank you for this very thorough explanation! 

I have one follow-up question here (which relates to the problem we ran into which made me open this issue):
If we consider a _warped_ grid source model - could `inwardshift` not potentially result in different numbers of grid points inside/outside? Meaning it should be done on the template if anywhere?

Regarding code and documentation:
- We could have a warning plus better documentation (docstring) of the `inwardshift`/ `moveinward` distinction regarding volume and surface grids.
- Depending on your feedback, documentation maybe as a FAQ for the warped grids could be beneficial.

I would be happy to have a stab at it!

**Comment by vlitvak** [comment]
Hi guys,

Just make sure not to break the interface as SPM relies on this
functionality.

Stay safe!

Vladimir

**Comment by robertoostenveld** [comment]
I would expect the MNI 3D grid template (in the different resoluitions) to already contain the Boolean `inside` specification. That does not change when warping/deforming positions. So the functionality related to `inwardshift` does not apply, since that is only for determining which positions are inside the volume conductor.

That is in this line https://github.com/fieldtrip/fieldtrip/blob/e7e2aa82cb291462f58f29b2dbc2c4352584775a/ft_prepare_sourcemodel.m#L687

```
% determine the dipole locations that are inside the source compartment of the
% volume conduction model, i.e. inside the brain
if ~isfield(sourcemodel, 'inside')
  sourcemodel.inside = ft_inside_headmodel(sourcemodel.pos, headmodel, 'grad', sens, 'headshape', cfg.headshape, 'inwardshift', cfg.inwardshift); % this returns a boolean vector
end
```

Just prior to that you see how the movewinward happens. Not particularly smart, it simply assumes that vertices on a sulcus that just peeps out of the brain compartment needs to be pushed inwards. This happens by projecting them (as if they were electrodes) on the surface. The surface here (line 672) is a slightly moved inward (with cfg.moveinward) version of the brain surface in the head model (which most likely is the only surface).

**Comment by schoffelen** [comment]
I agree with @robertoostenveld, although I can imagine that other users share Britta's expectation, i.e. that the inside is only determined after the warp, and by somehow magically coming up with the subject-specific boundary :), if not provided in cfg.headmodel. 

I guess that this would be a matter of managing the user expectations. 

Alternatively, one could consider the situation with cfg.warpmni = true, and in case a headmodel is also provided in the input cfg, the inside positions are determined as per the subject-specific headmodel

**Comment by robertoostenveld** [comment]
> I guess that this would be a matter of managing the user expectations.

We could give warnings if the user specifies something for these cfg options that is inconsistent with what the code does, i.e., warn if the option is specified but not used or not relevant.

**Comment by britta-wstnr** [comment]
It makes sense to me why `inwardshift` and warping is not a good combination, both on the computational level (i.e., the `inside/outside` is already defined in the template) but also the actual data analysis level (as it might result in a different number of `inside` voxels per subject which will prevent easy across-subjects analyses).

However, as reported above, we saw this behaviour:

> Having specified cfg.warpmni = 'yes' (and supplying the template etc.) makes cfg.inwardshift have no effect (which might be wanted as it would destroy the warped MNI structure on an individual basis?).
If also cfg.moveinward is specified the grid points get definitely moved - unless cfg.headmodel is not supplied, then the computation throws an error.

That makes me think: we could make sure that non-logical combinations throw a warning plus maybe a Q&A about warping and `inwardshift`?

**Comment by schoffelen** [comment]
Yes, it certainly makes sense. Please feel free to carry this forward.
